broadcaster/backlog/backlog.go

```diff
@@ -0,0 +1,317 @@
+// the given confirmed index
+func (b *backlog) removeFromLookup(start, end uint64) {
+	b.lookupLock.Lock()
+	defer b.lookupLock.Unlock()
+	for i := start; i == end; i++ {
+		delete(b.lookupByIndex, i)
+	}
+}
+
+func (b *backlog) lookup(i uint64) (*backlogSegment, error) {
+	b.lookupLock.RLock()
+	pointer, ok := b.lookupByIndex[i]
+	b.lookupLock.RUnlock()
+	if !ok {
+		return nil, fmt.Errorf("error finding backlog segment containing message with SequenceNumber %d", i)
+	}
+
+	s := pointer.Load()
+	if s == nil {
+		return nil, fmt.Errorf("error loading backlog segment containing message with SequenceNumber %d", i)
+	}
+
+	return s, nil
+}
+
+func (s *backlog) Count() int {
+	return int(s.messageCount.Load())
+}
+
+// reset removes all segments from the backlog
+func (b *backlog) reset() {
+	b.lookupLock.Lock()
+	defer b.lookupLock.Unlock()
+	b.head = atomic.Pointer[backlogSegment]{}
+	b.tail = atomic.Pointer[backlogSegment]{}
+	b.lookupByIndex = map[uint64]*atomic.Pointer[backlogSegment]{}
+	b.messageCount.Store(0)
+}
+
+type BacklogSegment interface {
+	Next() BacklogSegment
+	Messages() []*m.BroadcastFeedMessage
+}
+
+type backlogSegment struct {
+	start           atomic.Uint64
+	end             atomic.Uint64
+	messages        []*m.BroadcastFeedMessage
+	messageCount    atomic.Uint64
+	nextSegment     atomic.Pointer[backlogSegment]
+	previousSegment atomic.Pointer[backlogSegment]
+}
+
+func (s *backlogSegment) Next() BacklogSegment {
+	return s.nextSegment.Load()
+}
+
+func (s *backlogSegment) Messages() []*m.BroadcastFeedMessage {
+	return s.messages
+}
+
+// get reads messages from the given start to end MessageIndex
+func (s *backlogSegment) get(start, end uint64) ([]*m.BroadcastFeedMessage, error) {
+	noMsgs := []*m.BroadcastFeedMessage{}
+	if start < s.start.Load() {
+		return noMsgs, errOutOfBounds
+	}
+
+	if end > s.end.Load() {
+		return noMsgs, errOutOfBounds
+	}
+
+	startIndex := start - s.start.Load()
+	endIndex := end - s.start.Load() + 1
+	return s.messages[startIndex:endIndex], nil
+}
+
+// append appends the given BroadcastFeedMessage to messages if it is the first
+// message in the sequence or the next in the sequence. If segment's end
+// message is ahead of the given message append will do nothing. If the given
+// message is ahead of the segment's end message append will return
+// errDropSegments to ensure any messages before the given message are dropped.
+func (s *backlogSegment) append(prevMsgIdx uint64, msg *m.BroadcastFeedMessage) error {
+	seen := false
+	defer s.updateSegment(&seen)
+
+	if expSeqNum := prevMsgIdx + 1; prevMsgIdx == 0 || uint64(msg.SequenceNumber) == expSeqNum {
+		s.messages = append(s.messages, msg)
+	} else if uint64(msg.SequenceNumber) > expSeqNum {
+		s.messages = nil
+		s.messages = append(s.messages, msg)
+		return fmt.Errorf("new message sequence number (%d) is greater than the expected sequence number (%d): %w", msg.SequenceNumber, expSeqNum, errDropSegments)
```

Comment:
In this case, we need to not only clear this segment but also the rest of the buffer, to ensure sequence numbers are contiguous. We should also make sure to reset lookupByIndex.

Reply:
Yup, it returns the `errDropSegments` error which the `backlog.Append` method checks for. If it finds this error it will drop all previous segments. The current segment will be kept and it will start with this message.